jwt-bcp/draft-sheffer-oauth-jwt-bcp.md

```diff
@@ -97,89 +134,186 @@ updates that apply to this document.
+
+## Validate All Cryptographic Operations ## {#validate-crypto}
+
+All cryptographic operations used in the JWT MUST be validated and the entire JWT MUST be rejected
+if any of them fail to validate.
+This is true not only of JWTs with a single set of Header Parameters
+but also for Nested JWTs, in which both outer and inner operations MUST be validated
+using the keys and algorithms supplied by the application.
 
 <!-- See draft-ietf-oauth-jwsreq-13, sec. 6.1 and 6.2. -->
 
-## Distinguishing between Different Applications of JWTs
+## Validate Cryptographic Inputs ## {#validate-inputs}
+
+Some cryptographic operations, such as Elliptic Curve Diffie-Hellman key agreement
+("ECDH-ES") take inputs that may contain invalid values, such as points not on the specified elliptic curve
+or other invalid points.
+Either the JWS/JWE library itself must validate these inputs before using them
+or it must use underlying cryptographic libraries that do so (or both!).
+
+## Ensure Cryptographic Keys have Sufficient Entropy {#key-entropy}
+
+The Key Entropy and Random Values advice in Section 10.1 of [RFC7515] and
+the Password Considerations in Section 8.8 of [RFC7518]
+MUST be followed.
+
+## Use UTF-8 ## {#use-utf8}
 
+[RFC7515], [RFC7516], and [RFC7519] all specify that UTF-8 be used for encoding and decoding JSON
+used in Header Parameters and JWT Claims Sets.
+Implementations and applications MUST do this, and not use other Unicode encodings for these purposes.
+
+## Validate Issuer and Subject ## {#validate-iss-sub}
+
+When a JWT contains an "iss" (issuer) claim, the application MUST validate that the cryptographic keys
+used for the cryptographic operations in the JWT belong to the issuer.
+If they do not, the application MUST reject the JWT.
+
+Similarly, when the JWT contains a "sub" (subject) claim, the application MUST validate that
+the subject value corresponds to a valid subject and/or issuer/subject pair at the application.
+This may include confirming that the issuer is trusted by the application.
+If the issuer, subject, or the pair are invalid, the application MUST reject the JWT.
+
+## Use and Validate Audience ## {#use-aud}
+
+If the same issuer can issue JWTs that can be used by more than one relying party or application,
+the JWT MUST contain an "aud" (audience) claim that can be used to determine whether the JWT
+is being used by an intended party or was substituted by an attacker at an unintended party.
+Furthermore, the relying party or application MUST validate the audience value
+and if the audience value is not associated with the recipient, it MUST reject the JWT.
+
+## Use Mutually Exclusive Validation Rules for Different Kinds of JWTs ## {#preventing-confusion}
+
+<!-- I disagree with this text for several reasons.  For one, this probably won't work for existing applications.
+     It requires inventing a multiplicity of new MIME types that make semantic distinctions
```

Comment:
"typ" is a MIME type because James Manger convinced the JOSE WG that it should be. ;-)

I don't want to break existing deployments that are already secure.  I think that should be an explicit non-goal.  That said, it's fine to help make it easier for new deployments to avoid mistakes.  I view that as the point of the exercise.